lib/Mojo/Redis/Protocol.pm

```diff
@@ -0,0 +1,240 @@
+      _parse_debug(scalar => $curr) if DEBUG > 1;
+    }
+
+    # Arrays
+    elsif ($curr->{type} eq '*') {
+      $self->{stop} = index $$buf, "\r\n" if $self->{stop} < 0;
+      last CHUNK if $self->{stop} < 0;                   # Wait for more data
+
+      $curr->{len} = substr $$buf, 0, $self->{stop}, '';
+      last CHUNK if length($$buf) - 2 < $curr->{len};    # Wait for more data
+
+      $curr->{data} = $curr->{len} < 0 ? undef : [];
+      substr $$buf, 0, 2, '';                            # Remove \r\n after array size
+      _parse_debug(array => $curr) if DEBUG > 1;
+
+      # Fill the array with data
+      if ($curr->{len} > 0) {
+        $curr = $self->{curr} = {level => $curr->{level} + 1, parent => $curr};
+        next CHUNK;
+      }
+    }
+
+    # Should not come to this
+    else {
+      confess "[Mojo::Redis::Protocol] Unknown message type $curr->{type}";
+    }
+
+    # Wait for more data
+    next CHUNK unless exists $curr->{data};
+
+    # Fill parent array with data
+    while (my $parent = delete $curr->{parent}) {
+      push @{$parent->{data}}, $curr->{data};
+      _parse_debug(parent => $parent) if DEBUG > 1;
+
+      if (@{$parent->{data}} < $parent->{len}) {
+        $curr = $self->{curr} = {level => $curr->{level}, parent => $parent};
+        next CHUNK;
+      }
+      else {
+        $curr = $self->{curr} = $parent;
+      }
+    }
+
+    unless ($curr->{level}) {
+      push @messages, $curr;
+      _parse_debug(push => $curr) if DEBUG > 1;
+      $curr = $self->{curr} = {level => 0};
+    }
+  }
+
+  return @messages;
+}
+
+sub _parse_debug {
+  my ($type, $item) = @_;
+  local $_ = join ', ', map {"$_=$item->{$_}"} sort grep { !/^(level)$/ } keys %$item;
+  s!\r\n!\\r\\n!g;
+  warn "[Mojo::Redis::Protocol] ($item->{level}) $type - $_\n";
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojo::Redis::Protocol - Alternative to Protocol::Redis
+
+=head1 SYNOPSIS
+
+  use Mojo::Redis::Protocol;
+
+  my $protocol = Mojo::Redis::Protocol->new->encoding("UTF-8");
+
+  print $protocol->encode("foo");
+  print $protocol->encode({type => "-", data => "foo"});
+  print $protocol->encode(["foo", "bar", "baz"]);
+
+  while (my $msg = $protocol->parse($bytes)) {
+    ...
+  }
+
+=head1 DESCRIPTION
+
+L<Mojo::Redis::Protocol> is a Redis protocol parser -
+L<https://redis.io/topics/protocol>.
+
+The API is currently higly EXPERIMENTAL and should probably not be used
+directly by other projects. The reason for this is that L</encode>
+and L</decode> is not symmetrical.
+
+=head1 ATTRIBUTES
+
+=head2 encoding
+
+  $protocol = $protocol->encoding("UTF-8");
+  $protocol = $protocol->encoding(undef); # same as pass-through
+  $encoding = $protocol->encoding;
+
+Used to set or read the encoding.
+
+=head1 METHODS
+
+=head2 encode
+
+  $bytes = $protocol->encode($msg);
+
+Takes a message structure and turns it into bytes.
+
+=head2 new
+
+  $protocol = Mojo::Redis::Protocol->new;
+
+=head2 parse
```

Comment:
This also needs to be updated to match the current API (returns a list of messages)